src/network/networkd-dhcp4.c

```diff
@@ -575,6 +579,144 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
+        );
+
+        r = routing_policy_rule_get(link->manager, rule, &rrule);
+        if (r == 0) {
+            /* this means the rule already existed on the system, we just need to own it */
+            (void) routing_policy_rule_make_local(link->manager, rrule);
+        } else if (r < 0) {
+            /* this means we need to add the rule */
+            r = routing_policy_rule_configure(rule, link, dhcp4_routing_policy_rule_handler);
+            if (r < 0)
+                    return log_link_error_errno(link, r, "Failed to configure source routing policy rule");
+            link->dhcp4_messages++;
+        }
+
+        /* Swap over our internal representation of the rule to the new one we (possibly) added.
+        The dhcp4_remove_source_routing_policy_rule function will take care of actually calling
+        routing_policy_rule_remove based on the old lease address later on */
+        if (link->dhcp4_source_routing_policy_rule)
+                routing_policy_rule_free(TAKE_PTR(link->dhcp4_source_routing_policy_rule));
+        link->dhcp4_source_routing_policy_rule = TAKE_PTR(rule);
+        return 0;
+}
+
+static int dhcp4_remove_source_routing_policy_rule(Link *link, sd_dhcp_lease *lease, const struct in_addr *address, bool remove_all) {
+        _cleanup_(routing_policy_rule_freep) RoutingPolicyRule *rule = NULL;
+        _cleanup_free_ char *from_addr_str = NULL;
+        bool is_current_lease;
+        int r;
+
+        assert(link);
+        assert(address);
+
+        if (!link->network->dhcp_source_routing_enabled)
+                return 0;
+
+        r = routing_policy_rule_new(&rule);
+        if (r < 0)
+                return log_oom();
+
+        rule->from.in = *address;
+        rule->from_prefixlen = 32;
+        rule->family = AF_INET;
+        rule->priority = link->network->dhcp_source_routing_rule_priority;
+        if (link->network->dhcp_route_table_set)
+                rule->table = link->network->dhcp_route_table;
+
+        is_current_lease = link->dhcp4_source_routing_policy_rule &&
+            routing_policy_rule_compare_func(rule, link->dhcp4_source_routing_policy_rule) == 0;
+        if (!remove_all && is_current_lease)
+                return 0;
+
+        if (DEBUG_LOGGING) {
+                const union in_addr_union address_union = { .in = *address };
+                (void) in_addr_to_string(AF_INET, &address_union, &from_addr_str);
+        }
+        log_link_debug(
+                link, "Removing source routing rule %s/%u -> 0.0.0.0/0 table %d priority %d family %d",
+                from_addr_str,
+                rule->from_prefixlen,
+                rule->table,
+                rule->priority,
+                rule->family
+        );
+
+        (void) routing_policy_rule_remove(rule, link, NULL);
+        if (is_current_lease) {
+                routing_policy_rule_free(link->dhcp4_source_routing_policy_rule);
+                link->dhcp4_source_routing_policy_rule = NULL;
```

Comment:
hmm, we should rework routing_policy_rule_free() so that it returns `(RoutingPolicyRule*) NULL` unconditionally, so that we can write:

```c
link->dhcp4_source_routing_policy_rule = routing_policy_rule_free(link->dhcp4_source_routing_policy_rule);
```

we do things like that in almost all new code these days, and really should do this here too.

(of course, this is independent of this work of yours, but it would be very welcome to see a commit in your PR that converts this ;-))